File: shared/scripts/xcheck-merkle.py
# Temporary cross-language check: recomputes the RFC 6962 roots for n = 1…8
# with hashlib and diffs them against shared/scripts/xcheck-merkle.mjs.
import hashlib
import json
import subprocess
import sys
from pathlib import Path

mjs = Path(__file__).resolve().parent / "xcheck-merkle.mjs"

def sha256(b):
    return hashlib.sha256(b).digest()


def mth(leaves):
    if len(leaves) == 1:
        return sha256(b"\x00" + leaves[0])
    k = 1
    while k * 2 < len(leaves):
        k *= 2
    return sha256(b"\x01" + mth(leaves[:k]) + mth(leaves[k:]))


reference = {}
for n in range(1, 9):
    leaves = sorted(sha256(f"head-{i}".encode()) for i in range(n))
    reference[f"n{n}"] = mth(leaves).hex()

result = subprocess.run(
    ["node", str(mjs)],
    capture_output=True, text=True, check=True,
)
js = json.loads(result.stdout)

failures = 0
for name, expected in reference.items():
    actual = js.get(name)
    if actual == expected:
        print(f"PASS {name}")
    else:
        failures += 1
        print(f"FAIL {name}\n  python: {expected}\n  js:     {actual}")

print(f"\n{len(reference) - failures}/{len(reference)} roots identical")
sys.exit(1 if failures else 0)

File: shared/scripts/xcheck-merkle.mjs
// Temporary cross-language check: prints merkle roots for n = 1…8 over
// deterministic 32-byte leaves so it can be diffed against Python hashlib.
// Not part of the test suite.
import { merkleRoot, bytesToHex } from "../merkle.js";

async function sha256(bytes) {
  return new Uint8Array(await globalThis.crypto.subtle.digest("SHA-256", bytes));
}

const leaf = async (i) => {
  // Mirrors the Python side: leaf_i = SHA-256(utf8("head-" + i))
  return sha256(new TextEncoder().encode(`head-${i}`));
};

const out = {};
for (let n = 1; n <= 8; n++) {
  const leaves = [];
  for (let i = 0; i < n; i++) leaves.push(await leaf(i));
  leaves.sort((a, b) => (bytesToHex(a) < bytesToHex(b) ? -1 : 1));
  out[`n${n}`] = bytesToHex(await merkleRoot(leaves));
}
console.log(JSON.stringify(out, null, 2));